### ISSUE TITLE
Doesn't infer type of default parameters

### ISSUE THREAD

**Issue by jteplitz** [palantir/tslint#646]
If you run tslint with requireParameterType set to true, you will get an error on the following function definition:

``` TypeScript
function foo(bar = true)
```

However TypeScript automatically infers the signature of this function to be 

``` TypeScript
(bar: boolean)
```

Tslint should infer the types of default parameters so that you don't have to write redundant code such as

``` TypeScript
function foo(bar: boolean = true)
```

**Comment by JKillian** [comment]
@jteplitz602 I think the rule's intent is about forcing types to be explicitly declared, even in situations like these. For example, the return type of functions can often be inferred, but this rule still requires type annotations on them. What do you think?

**Comment by jteplitz602** [comment]
@JKillian That sounds reasonable. I suppose forcing explicitness is a good thing.

**Comment by Fraegle** [comment]
I think explicit type declarations only make sense, where they can't be inferred. I would love to have an option on the rule to accept inferred types as type declarations.

**Comment by JKillian** [comment]
@Fraegle If you're working in an editor that doesn't show type information, it can make sense to have explicit type declarations. It can also add to code scanability at times. 

What you're asking for seems reasonable, but isn't it essentially what you get with the typescript compiler and disallowing implicit `any` declarations?

**Comment by andy-hanson** [comment]
`requireParameterType` seems to have been removed at some point?

**Comment by JoshuaKGoldberg** [comment]
Yes, this seems to have aged away.
